module/ice-9/generic-hash-tables.scm

```diff
@@ -0,0 +1,854 @@
+  (case (ht-weakness ht)
+    ((#f weak-value) #f)
+    ((weak-key weak-key-and-value) 'weak-keys)))
+
+(define (hash-table-value-weakness ht)
+  "Returns WEAK-VALUES if HT has weak values, or #f otherwise."
+  ;; If Guile ever supports ephemeral values, this procedure should
+  ;; return EPHEMERAL-VALUES if the HT values are ephemeral.
+  (case (ht-weakness ht)
+    ((#f weak-key) #f)
+    ((weak-value weak-key-and-value) 'weak-values)))
+
+(define (hash-table-weakness ht)
+  "Return the weakness of HT according to SRFI-126 spec."
+  (ht-weakness ht))
+
+;; This internal function allows us to create immutable hash tables
+(define (%make-hash-table equiv-function hash-function mutable capacity weakness)
+  (receive (internal-hash-function hash-function)
+      (get-hash-functions equiv-function hash-function)
+    (let ((real-table ((guile-ht-ctor weakness) capacity)))
+      ;; Arguments: real-table hash-function associator
+      ;;            weakness mutable? equivalence-function orig-hash-function
+      (make-generic-hash-table real-table internal-hash-function
+                               (equivalence-proc->associator equiv-function)
+                               weakness (and mutable #t)
+                               equiv-function hash-function))))
+
+;; If the list of arguments is updated, HASH-TABLE, ALIST->HASH-TABLE,
+;; HASH-TABLE-UNFOLD and HASH-TABLE-MAP should be updated as well.
+(define* (make-hash-table equiv-function hash-function
+                          #:key (capacity 1) (weakness #f) #:rest args)
+  "Creates a new hash table. EQUIV-FUNCTION is used as the comparison
+function and HASH-FUNCTION, if not false, is used as the hash function,
+otherwise a suitable hash-function for the hash table is guessed or, if
+one can't be guessed, an error is signaled. #:WEAKNESS should be either
+#f, WEAK-KEY, WEAK-VALUE or WEAK-KEY-AND-VALUE, as specified by
+SRFI-126. #:CAPACITY is the minimal number of buckets of the hash table.
+
+ARGS is ignored and reserved for future extensions."
+  (%make-hash-table equiv-function hash-function #t capacity weakness))
+
+;; We use this to denote missing or unspecified values to avoid
+;; possible collision with *unspecified*.
+
+(define ht-unspecified (cons *unspecified* "ht-value"))
+
+(define* (hash-table equiv-function-or-mht-args . args)
+  "Creates a new immutable hash table with the associations given by
+ARGS. If EQUIV-FUNCTION-OR-MHT-ARGS is a list, the new hash table is
+created by (APPLY MAKE-HASH-TABLE EQUIV-FUNCTION-OR-MHT-ARGS), otherwise
+it is created by (MAKE-HASH-TABLE EQUIV-FUNCTION-OR-MHT-ARGS) with the
+initial capacity set to the number of associations in args.
+
+The ARGS alternate between keys and values. If the same key (in the
+sense of the equality procedure) is specified more then once, an error
+is signaled."
+  (let ((len (length args)))
+    (unless (even? len)
+      (error "Odd number of key-value pairs" args))
+    (let* ((capacity (quotient len 2))
```

Comment:
`let*` is not required here.

Reply:
I removed all useless `let*` in that file.